File: backend/simulator.ts
/** Educational replay engine using deterministic synthetic prices and fictitious index units.
 * It never imports a broker or places orders. Illustrative costs omit exchange taxes and real
 * contract rules, so results must not be interpreted as an exchange backtest or expected profit.
 */
import type { SymbolCode, Trade, ReplayResult } from "./types.js";

/** Generate sample prices, apply an EMA crossover, and return fills/equity after a final close.
 * A fixed random seed makes tests reproducible. Positions cannot spend more than available cash.
 * A 2% sample-equity loss stops new entries; it is not a live-market risk-control implementation.
 */
export function simulateSyntheticStrategy(
  symbol: SymbolCode,
  capital: number,
  fast: number,
  slow: number,
): ReplayResult {
  let randomSeed = 42;
  /** Produce a deterministic pseudo-random value for sample-price noise, never for security. */
  const nextRandomValue = () => {
    randomSeed = (Math.imul(randomSeed, 1664525) + 1013904223) >>> 0;
    return randomSeed / 4294967296;
  };
  /** Round display/fill amounts to two decimal places for this educational model. */
  const roundCurrency = (amount: number) => Math.round(amount * 100) / 100;
  const basePrice = { NIFTY: 24000, BANKNIFTY: 51000, SENSEX: 79000 }[symbol];
  const prices = Array.from({ length: 240 }, (_, bar) =>
    roundCurrency(
      basePrice * (1 + 0.006 * Math.sin(bar / 12) + 0.000025 * bar) +
        nextRandomValue() * 20 -
        10,
    ),
  );
  let fastEma = prices[0],
    slowEma = prices[0],
    positionQuantity = 0,
    entryPrice = 0;
  let availableCash = capital,
    equityPeak = capital,
    maximumDrawdown = 0,
    riskStopped = false;
  const trades: Trade[] = [],
    equity: number[] = [];
  prices.forEach((price, bar) => {
    fastEma += ((price - fastEma) * 2) / (fast + 1);
    slowEma += ((price - slowEma) * 2) / (slow + 1);
    if (availableCash + positionQuantity * price <= capital * 0.98) {
      riskStopped = true;
    }
    if (
      positionQuantity &&
      (fastEma < slowEma || riskStopped || bar === prices.length - 1)
    ) {
      const exitPrice = roundCurrency(price * 0.9995),
        proceeds = positionQuantity * exitPrice - 5;
      trades.push({
        bar,
        side: "SELL",
        price: exitPrice,
        quantity: positionQuantity,
        pnl: roundCurrency(proceeds - positionQuantity * entryPrice - 5),
      });
      availableCash += proceeds;
      positionQuantity = 0;
    } else if (
      !positionQuantity &&
      bar >= slow &&
      bar < prices.length - 1 &&
      fastEma > slowEma &&
      !riskStopped
    ) {
      entryPrice = roundCurrency(price * 1.0005);
      positionQuantity = Math.max(
        0,
        Math.floor((availableCash - 5) / entryPrice),
      );
      if (positionQuantity) {
        availableCash -= positionQuantity * entryPrice + 5;
        trades.push({
          bar,
          side: "BUY",
          price: entryPrice,
          quantity: positionQuantity,
          pnl: null,
        });
      }
    }
    const portfolioValue = availableCash + positionQuantity * price;
    equityPeak = Math.max(equityPeak, portfolioValue);
    maximumDrawdown = Math.max(maximumDrawdown, equityPeak - portfolioValue);
    equity.push(roundCurrency(portfolioValue - capital));
  });
  return {
    pnl: roundCurrency(availableCash - capital),
    drawdown: roundCurrency(maximumDrawdown),
    equity,
    trades,
    bars: prices.length,
    source: "synthetic",
    risk_stopped: riskStopped,
    cost_model: "0.05% slippage per fill + ₹5 per order; taxes excluded",
  };
}

File: backend/simulator.py
"""Deterministic educational replay. Synthetic index units, not exchange contracts."""
import math
import random


def replay(symbol, capital, fast, slow):
    rng = random.Random(42)
    base = {'NIFTY': 24000, 'BANKNIFTY': 51000, 'SENSEX': 79000}[symbol]
    prices = [round(base * (1 + .006 * math.sin(i / 12) + .000025 * i) + rng.uniform(-10, 10), 2) for i in range(240)]
    fast_ema = slow_ema = prices[0]
    qty = 0
    entry = 0
    cash = float(capital)
    trades, equity = [], []
    peak = capital
    drawdown = 0
    stopped = False
    for i, price in enumerate(prices):
        fast_ema += (price - fast_ema) * 2 / (fast + 1)
        slow_ema += (price - slow_ema) * 2 / (slow + 1)
        value = cash + qty * price
        if value <= capital * .98:
            stopped = True
        if qty and (fast_ema < slow_ema or stopped or i == len(prices) - 1):
            fill = round(price * .9995, 2)
            proceeds = qty * fill - 5
            pnl = round(proceeds - qty * entry - 5, 2)
            trades.append({'bar': i, 'side': 'SELL', 'price': fill, 'quantity': qty, 'pnl': pnl})
            cash += proceeds
            qty = 0
        elif not qty and i >= slow and i < len(prices) - 1 and fast_ema > slow_ema and not stopped:
            entry = round(price * 1.0005, 2)
            qty = max(0, int((cash - 5) // entry))
            if qty:
                cash -= qty * entry + 5
                trades.append({'bar': i, 'side': 'BUY', 'price': entry, 'quantity': qty, 'pnl': None})
        value = cash + qty * price
        peak = max(peak, value)
        drawdown = max(drawdown, peak - value)
        equity.append(round(value - capital, 2))
    return {'pnl': round(cash - capital, 2), 'drawdown': round(drawdown, 2), 'equity': equity,
            'trades': trades, 'bars': len(prices), 'source': 'synthetic', 'risk_stopped': stopped,
            'cost_model': '0.05% slippage per fill + ₹5 per order; taxes excluded'}
